### PR TITLE
Multistage ONBUILD COPY Support

### LINKED ISSUE #533: ONBUILD COPY fails using multi-stage build
Hi everyone, thanks for your work on Kaniko!

**Actual behavior**

When using `ONBUILD` instructions with a multi-stage build, Kaniko fails to copy files between stages:

```
+ docker run --rm -it --volume /Volumes/Users/mbfisher/git-projects/kaniko-multi-stage-bug:/build gcr.io/kaniko-project/executor:debug --context=/build --destination=my-kaniko-bug
INFO[0000] Downloading base image mbfisher/kaniko-multi-stage-bug:builder
2019/01/17 10:05:39 No matching credentials were found, falling back on anonymous
INFO[0001] Executing 1 build triggers
INFO[0001] Unpacking rootfs as cmd RUN mkdir /app && echo "Hello, World!" > /app/index requires it.
INFO[0001] Taking snapshot of full filesystem...
INFO[0001] Skipping paths under /kaniko, as it is a whitelisted directory
INFO[0001] Skipping paths under /build, as it is a whitelisted directory
INFO[0001] Skipping paths under /proc, as it is a whitelisted directory
INFO[0001] Skipping paths under /dev, as it is a whitelisted directory
INFO[0001] Skipping paths under /sys, as it is a whitelisted directory
INFO[0001] Skipping paths under /busybox, as it is a whitelisted directory
INFO[0001] RUN mkdir /app && echo "Hello, World!" > /app/index
INFO[0001] cmd: /bin/sh
INFO[0001] args: [-c mkdir /app && echo "Hello, World!" > /app/index]
INFO[0001] Taking snapshot of full filesystem...
INFO[0001] Skipping paths under /kaniko, as it is a whitelisted directory
INFO[0001] Skipping paths under /build, as it is a whitelisted directory
INFO[0001] Skipping paths under /proc, as it is a whitelisted directory
INFO[0001] Skipping paths under /dev, as it is a whitelisted directory
INFO[0001] Skipping paths under /sys, as it is a whitelisted directory
INFO[0001] Skipping paths under /busybox, as it is a whitelisted directory
INFO[0001] Deleting filesystem...
INFO[0002] Downloading base image mbfisher/kaniko-multi-stage-bug:runtime
2019/01/17 10:05:41 No matching credentials were found, falling back on anonymous
INFO[0002] Executing 1 build triggers
INFO[0002] Taking snapshot of full filesystem...
INFO[0002] Skipping paths under /kaniko, as it is a whitelisted directory
INFO[0002] Skipping paths under /build, as it is a whitelisted directory
INFO[0002] Skipping paths under /proc, as it is a whitelisted directory
INFO[0002] Skipping paths under /dev, as it is a whitelisted directory
INFO[0002] Skipping paths under /sys, as it is a whitelisted directory
INFO[0002] Skipping paths under /busybox, as it is a whitelisted directory
INFO[0002] COPY --from=builder /app/* /app/
error building image: error building stage: lstat /kaniko/builder: no such file or directory
```

**Expected behavior**
The build should succeed, as it does when using the Docker Engine:

```
+ docker build --no-cache -t my-multi-stage-build .
Sending build context to Docker daemon  50.69kB
Step 1/2 : FROM mbfisher/kaniko-multi-stage-bug:builder AS builder
# Executing 1 build trigger
 ---> Running in c6bbc064f397
Removing intermediate container c6bbc064f397
 ---> c3d2e2f359f7
Step 2/2 : FROM mbfisher/kaniko-multi-stage-bug:runtime
# Executing 1 build trigger
 ---> 2b4f93925947
Successfully built 2b4f93925947
Successfully tagged my-multi-stage-build:latest
+ docker run --rm my-multi-stage-build
Hello, World!
```

**To Reproduce**

See https://github.com/mbfisher/kaniko-multi-stage-bug

**Additional Information**
 - Possibly related to https://github.com/GoogleContainerTools/kaniko/issues/460?

### PR DESCRIPTION
**Pull request by dani29**
Fixes #533.

**Description**

ONBUILD did not work correctly, because the `resolveOnBuild()` method occurred after the `resolveStages()` method was called, so commands like `COPY --from=builder` were not properly replaced with `--from=0`. Moreover, because the `CalculateDependency()` method were unaware of these dependencies, files copied onbuild weren't even registered to be kept beyond the stage they were used it.

I had to break down `dockerfile.Stages(opts)`, which was a huge method, into smaller chunks doing the following:
1. parsing the dockerfile into []instructions.Stage (no changes here)
2. resolving instructions within stages with references to previous stages (keeping a mapping of stageName:stageIndex)
3. building KanikoStages from the []instructions.Stage (no changes here)

afterwards, adding the following:
- passing the map from step 2 to CalculateDependency() so it would register dependencies for ONBUILD files
- passing the map from step 2 to NewStageBuilder -> resolveOnBuild(), so the COPY commands "From" will be replaced with the stage index.
 
**Submitter Checklist**

These are the criteria that every PR should meet, please check them off as you
review them:

- [X] Includes [unit tests](../DEVELOPMENT.md#creating-a-pr)
- [ ] Adds integration tests if needed.

_See [the contribution guide](../CONTRIBUTING.md) for more details._

**Reviewer Notes**

- [ ] The code flow looks good. 
- [ ] Unit tests and or integration tests added.

**Release Notes**

Describe any changes here so maintainer can include it in the release notes, or delete this block.

```
Examples of user facing changes:
- Support ONBUILD Copy in multistage builds (#533)

```

### CODE REVIEW

**Comment by tejal29** [comment]
@dani29 the code flows looks good. 
Thanks for the refactoring!

File: pkg/executor/build.go [hunk 1]
```diff
@@ -52,6 +52,11 @@ import (
 // This is the size of an empty tar in Go
 const emptyTarSize = 1024
 
+// for testing
+var (
+	InitializeConfig = initializeConfig
```
Comment (tejal29) [inline pkg/executor/build.go:57]: Please make this a local variable so only the tests in `build` package can override this.

Comment (dani29) [reply pkg/executor/build.go:57]: 👍

**Comment by dani29** [comment]
Yes, I've verified that the Dockerfile provided by @mbfisher in #533 can be built.
The other docker file did not have the same issue with ONBUILD.

**Comment by dani29** [comment]
@JordanGoasdoue, I've had to manually fix conflicts since git wasn't able to digest the changes of both this PR and #1160. 
Any chance you could take a quick look and make sure I mess up anything with your flow?

**Comment by JordanGoasdoue** [comment]
Done @dani29 , it's all good

**Comment by kirkvogen** [comment]
Thank you for making this fix. I’ve been playing around with Kaniko and ran into this issue today. And then I do a search and find that you fixed it 15 minutes ago. Serendipity!

**Comment by kirkvogen** [comment]
I tried the latest releases of the image. It fixed the issue that I was seeing with multi-stage ONBUILD. Thanks!

**Pull request merged**